web_external/js/views/map/MapAdapters.js

```diff
@@ -0,0 +1,173 @@
+                        this.trigger('m:map_adapter_layerError', layer);
+                    }, this).initLayer(mapContainer, dataset, dataset.get('geoData'), visProperties);
+                }, this).loadGeoData();
+                //
+                // Instead of dataset.loadGeoData, ideally something like
+                // girder.RestRequest({
+                //     type: "POST"
+                //     url: "adapter/" + dataset._id + "/" + layerType.toString()(),
+                //     params: userInput,
+                // }).success(function (data){
+                //     createLayer
+                // });
+            }
+        };
+    }
+    minerva.adapters.MapAdapterRegistry = _.extend(new MapAdapterRegistry(), Backbone.Events);
+})();
+
+minerva.representations = minerva.representations || {};
+minerva.representations.defineMapLayer = function (layerType, layerDefinition, parentDefinition) {
+    if (parentDefinition) {
+        layerDefinition.prototype = new parentDefinition();
+    }
+    minerva.adapters.MapAdapterRegistry.register(layerType, layerDefinition);
+    return layerDefinition;
+}
+
+minerva.representations.MapLayer = minerva.representations.defineMapLayer('map', function () {
+    this.deleteLayer = function (mapContainer) {
+        mapContainer.deleteGeoJsLayer(this.geoJsLayer);
+    },
+
+    this.setOpacity = function (opacity) {
+        this.geoJsLayer.opacity(opacity);
+    },
+
+    this.collectUserInput = function (dataset) {
+        return {};
+    }
+});
+
+minerva.representations.JsonReaderMapLayer = minerva.representations.defineMapLayer('geojson', function () {
+    this.readerType = 'jsonReader',
+
+    this.initLayer = function (mapContainer, dataset, data, visProperties) {
+        this.geoJsLayer = mapContainer.createGeoJsLayer('feature');
+        try {
+            var reader = geo.createFileReader(this.readerType, {layer: this.geoJsLayer});
+            reader.read(data, _.bind(function () {
+                this.trigger('m:map_layer_renderable', this);
+            }, this));
+        } catch (err) {
+            console.error('This layer cannot be rendered to the map');
+            console.error(err);
+            this.trigger('m:map_layer_error', this);
+        }
+    }
+}, minerva.representations.MapLayer);
+
+minerva.representations.ContourJsonReaderMapLayer = minerva.representations.defineMapLayer('contour', function () {
+    this.readerType = 'contourJsonReader'
+}, minerva.representations.JsonReaderMapLayer);
+
+minerva.representations.ChoroplethMapLayer = minerva.representations.defineMapLayer('choropleth', function () {
+    this.collectUserInput = function (dataset) {
+        return {
+            colorByValue: dataset.getMinervaMetadata().colorByValue,
+            colorScheme: dataset.getMinervaMetadata().colorScheme
+        };
+    },
+
+    this.initLayer = function (mapContainer, dataset, jsonData, visProperties) {
```

Comment:
My thinking was that this allows us to decouple the dataset from the data itself.  This API is obviously a work in progress and I'm trying to figure out the right approach.

So when you call initLayer, you pass in the dataset which can be used in a callback, for instance a click info widget that needs to use the dataset to get to some details.  

The adapter takes the dataset along with other properties and produces the jsonData, which is the actual data that should be rendered.  With a different (x,y) center, zoom, or some other properties set, the adapter can get different jsonData for the same dataset.  The dataset isn't responsible for pushing the displayed data to the layer, so it is decoupled.  

For now, we actually get the data from the dataset (in a process mediated by the Adapter Registry), but with this API, it doesn't have to be that way and we could get the data from a the server via an adapter.